### ISSUE TITLE
Best way to find all hexes with a certain radius of given hex

### ISSUE THREAD

**Issue by derimmer** [vasl-developers/vasl#1895]
Need some coding advice . . . 

To have the LOS engine work with FFE, and in particular, NOBA Bombardments using Smoke, I would like to find all of hexes within a given range of the placement hex of the counter. 

(1) For a normal-sized FFE, this is quite easy as we can go around the clock to identify the six adjacent hexes and I have already created a method to do so. However at a range of 2 and 3 hexes, that methodology starts to get a bit cumbersome although still possible. I just think there might be a better way. . . . 

(2) One method would be to loop through all of the hexes of the map, calculate their range from the origin hex and identify those within the desired range or less. This seemed to be quite simple and accurate.  It would be easy to calculate the rows and columns that could possibly contain such hexes and limit the testing to only hexes within those bounds.  

The difficulty comes in designing a for . . . loop or for each . . . loop in java. We make great use of nested loops in the los code but they can be tricky to manage depending on crop and flip configurations. On standard geo boards, we have 10-hex columns (A, C, E. . . . ) and 11-hex columns (B, D, F. . . . ). I would rather avoid this approach. 

Is there a for . . . each function in java that will work with a 2 dimensional array of different lengths?

(3) Is there another way of crunching this that I have not yet thought of? This seems like the most likely path forward. LOL. Any ideas?

NB: Any method must be able to contend with null results (if for example, testing for a radius value of 3 when the origin hex is two hexes from a board edge - as can easily happen). 

TIA for any pointers, advice, or solutions.

**Comment by geezer09** [comment]
Hey Doug, I haven’t taken a look at how overlays are handled in terms of LOS, but my initial thought is the following:

From my understanding, the los tool looks at individual pixels to determine results, so why don’t we simply make the overlays replace the pixels of the original map so that whatever pixels are used on the overlays are used to determine los results, in this way you don’t have to custom code every single overlay to make it work with LOS code. 

Perhaps this doesn't work with terrain/overlays that changes during the game?  

If thats the case perhaps it should be the responsibility of these dynamic overlays to “mark” the map so that the los code doesn’t have to make these complicated and error prone determinations of whether a hex contains a dynamic overlay or not.

**Comment by derimmer** [comment]
Thanks for these comments. See mine below.

> Hey Doug, I haven’t taken a look at how overlays are handled in terms of LOS, but my initial thought is the following:
> 
> From my understanding, the los tool looks at individual pixels to determine results, so why don’t we simply make the overlays replace the pixels of the original map so that whatever pixels are used on the overlays are used to determine los results, in this way you don’t have to custom code every single overlay to make it work with LOS code.
> 
This is exactly the work I have been doing over the past several years with respect to "traditional" overlays added via boardPicker. Given the stability of these overlays once placed on the map, this is absolutely the right approach. While my code needs some refactoring, it does seem to work.  It may work for the drag-and-drop overlays from the overlay extension. My mind is not made up one way or another yet.
>
> Perhaps this doesn't work with terrain/overlays that changes during the game?
>
There is no perhaps in my mind. It would be very difficult as you would have to manage every minor adjustment to such counters, put back the original data, etc, etc. In the case of OBA counters, you would have to both keep the underlying terrain information and add to it the FFE impacts. For the draggable overlays on the toolbar, this would be a nightmare.  
> 
> If thats the case perhaps it should be the responsibility of these dynamic overlays to “mark” the map so that the los code doesn’t have to make these complicated and error prone determinations of whether a hex contains a dynamic overlay or not.
>
This determination is only relevant to draggable overlays in which the image extends beyond a single hex. OBA being the prime example. But I agree that this could be driven by the overlays when they are placed. In fact, in many instances that is what happens. However and whenever it is done, it would still require knowing which hexes should be "marked", no?

Thanks!

**Comment by geezer09** [comment]
Ok thanks for the info. I guess i didn’t really answer any of your questions. 

As for 2), to iterate through a 2d array is fairly straightforward:

        int[][] array2D = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9}
        };

        
        for (int[] row : array2D) { 
            for (int element : row) { 
                System.out.print(element + " "); 
}
}

Is there a nuance in your question that I am missing?

Edit:

As for your question for 1), cant you just recursively use the same methodology you used for finding hexes that are adjacent to find those that are 2 hexes away? Whatever function you used to locate the adjacent hexes, call it on those hexes to find those that are two hexes away, eliminating the duplicates that are returned.

**Comment by derimmer** [comment]
> As for your question for 1), cant you just recursively use the same methodology you used for finding hexes that are adjacent to find those that are 2 hexes away? Whatever function you used to locate the adjacent hexes, call it on those hexes to find those that are two hexes away, eliminating the duplicates that are returned.
>

Yes, I certainly could. It felt like a brute force solution. I was hoping something more elegantly efficient! 

I see your example of iterating through a 2darray. I was trying something similar to iterate through HexGrid[][] and was having problems. Probably over-thinking it or just scrambling the syntax. 

I will try it again first. And then resort to brute force. 

Thanks

**Comment by zgrose** [comment]
If you haven't tried yet, ChatGPT/etc can often be helpful if you are 
A) starting with something and you want to tweak it, or
B) you can describe what you want but you need some help implementing it.

It often explains, or you can ask it to explain, why it did what it did. Helps a lot with learning new tricks/concepts.

Super handy when you're pretty sure you're doing something a lot of people have already done.

**Comment by uckelman** [comment]
If you're looking for mathematical operations on a hex grid, look here: https://www.redblobgames.com/grids/hexagons/ Specifically what you want is under "Neighbors".

**Comment by derimmer** [comment]
Thanks to all for your comments. @uckelman I have checked out redblobgames. A great resource if somewhat overwhelming to my non-math trained brain!

The following code loop appears to be working:

```
public String [] getAllHexesInRadiusOf(Hex nobaHex, double nobaradius) {
        int maxhexes = (int) ((nobaradius/2) * (1 + nobaradius) *6) +1;
        int x = 0;
        String[] hexes = new String[maxhexes];
        for (Hex[] hexrow : hexGrid) {
            for (Hex testcol : hexrow) {
                if (range (testcol, nobaHex, getMapConfiguration()) <= nobaradius) {
                    hexes[x] = testcol.getName();
                    x += 1;
                }
            }
        }
        return hexes;
    }
```

The advantage of this method is that it avoids testing against null hexes. My initial problem was with writing the two "for . . ." lines. I could not get my head around the syntax. But thanks to @geezer09 and his reply above, I finally saw the light.

**Comment by geezer09** [comment]
Glad you figured it out.

Not sure where that function is located but it is general enough that it could make sense to place it in the map or board class and make sure-it can be called by anyone that needs to solve the problem of finding hexes within a range of a hex. The name of the parameters you chose are very case specific.

I’m surprised it hasn’t been written already.

**Comment by derimmer** [comment]
> Glad you figured it out.
> 
> Not sure where that function is located but it is general enough that it could make sense to place it in the map or board class and make sure-it can be called by anyone that needs to solve the problem of finding hexes within a range of a hex. The name of the parameters you chose are very case specific.
> 
> I’m surprised it hasn’t been written already.

It is in the LOS.Map.Map class along with most of the other LOS code. The HexGrid involved is los specific and is not used for other purposes so I think this is the best place for it. 

You are correct about the parameters and I will change them to make it a generic method.

**Comment by uckelman** [comment]
> Thanks to all for your comments. @uckelman I have checked out redblobgames. A great resource if somewhat overwhelming to my non-math trained brain!
> 
> The following code loop appears to be working:
> 
> ```
> public String [] getAllHexesInRadiusOf(Hex nobaHex, double nobaradius) {
>         int maxhexes = (int) ((nobaradius/2) * (1 + nobaradius) *6) +1;
>         int x = 0;
>         String[] hexes = new String[maxhexes];
>         for (Hex[] hexrow : hexGrid) {
>             for (Hex testcol : hexrow) {
>                 if (range (testcol, nobaHex, getMapConfiguration()) <= nobaradius) {
>                     hexes[x] = testcol.getName();
>                     x += 1;
>                 }
>             }
>         }
>         return hexes;
>     }
> ```

This is extremely inefficient, as you're looping over every hex in the grid, when you could instead directly compute which ones you need.

**Comment by derimmer** [comment]
> This is extremely inefficient, as you're looping over every hex in the grid, when you could instead directly compute which ones you need.

By "directly compute which ones you need", I assume you are referring to the methodology found on redblobgames. If I understood that methodology, it is essentially what @geezer09 suggested:

> cant you just recursively use the same methodology you used for finding hexes that are adjacent to find those that are 2 hexes away? Whatever function you used to locate the adjacent hexes, call it on those hexes to find those that are two hexes away, eliminating the duplicates that are returned.

I will try to construct a method that does this.

**Comment by geezer09** [comment]
As previously mentioned, ChatGPT is your friend here. I didn't realize what a beast of a class the weirdly named LOS.map.map is, but it already has a find adjacent hex method. Assuming it works okay and is not terribly inefficient, give it to GPT and ask for a function to return hexes within range of a certain hex and itll give you the following:


  
    public List<Hex> getHexesWithinDistance(Hex center, int distance) {
        // Keep track of all the hexes we've already encountered,
        // so we don't process the same hex multiple times.
        Set<Hex> visited = new HashSet<>();
        // We'll use a queue to do a BFS around the center hex.
        Queue<Hex> frontier = new LinkedList<>();
        // A map to store the distance of each hex from the center.
        Map<Hex, Integer> distanceMap = new HashMap<>();
  
      // Initialize the BFS with the center hex.
      visited.add(center);
      frontier.add(center);
      distanceMap.put(center, 0);
  
      // Perform the BFS until there are no more neighbors to explore.
      while (!frontier.isEmpty()) {
          Hex current = frontier.poll();
          int currentDistance = distanceMap.get(current);
  
          // Only expand if we have not reached the desired distance yet.
          if (currentDistance < distance) {
              // Explore all 6 neighboring hexes.
              for (int side = 0; side < 6; side++) {
                  Hex neighbor = getAdjacentHex(current, side);
                  // Make sure neighbor is valid and hasn't been visited before
                  if (neighbor != null && !visited.contains(neighbor)) {
                      visited.add(neighbor);
                      frontier.add(neighbor);
                      distanceMap.put(neighbor, currentDistance + 1);
                  }
              }
          }
      }
  
      // Return all visited hexes (those are within the given distance).
      return new ArrayList<>(visited);
  }

As @uckelman said, doing a direct range computation would be faster but since we are only searching for 10 or so hexes, I don't think that will make a noticable differece.

**Comment by derimmer** [comment]
In parallel with @geezer09's ChatGPT solution, I came up with the following:

```
public String [] getAllHexesInRadiusOf(Hex originHex, double arrayradius) {
        int maxhexes = (int) ((arrayradius / 2) * (1 + arrayradius) * 6) + 1;
        Hex startHex;
        int x = 0;
        String[] hexes = new String[maxhexes];
        hexes[x] = originHex.getName();
        // recursive loop
        for (String hexname : hexes) {
            startHex = getHex(hexname);
            String[] adjacentHexes = getAllAdjacentHexes(startHex);
            List hexesList = Arrays.asList(hexes);
            for (String adjhexname : adjacentHexes) {
                if (!hexesList.contains(adjhexname) && (range (getHex(adjhexname), originHex, getMapConfiguration()) <= arrayradius)) {
                    x += 1;
                    hexes[x] = adjhexname;
                }
            }
        }
        return hexes;
    }
```

It appears to work.
